compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -446,12 +447,195 @@ impl<'a> Parser<'a> {
+    pub(super) fn maybe_emit_stashed_expr_in_pats(&mut self, stmt: &Stmt) {
+        if self.dcx().has_errors().is_none() {
+            // No need to walk the statement if there's no stashed errors.
+            return;
+        }
+
+        struct PatVisitor<'a> {
+            /// `self`
+            parser: &'a Parser<'a>,
+            /// The freshly-parsed statement.
+            stmt: &'a Stmt,
+            /// The current match arm (for arm guard suggestions).
+            arm: Option<&'a Arm>,
+            /// The current struct field (for variable name suggestions).
+            field: Option<&'a PatField>,
+        }
+
+        impl<'a> PatVisitor<'a> {
+            /// Looks for stashed [`StashKey::ExprInPat`] errors in `stash_span`, and emit them with suggestions.
+            /// `stash_span` is contained in `expr_span`, the latter being larger in borrow patterns;
+            /// ```txt
+            /// &mut x.y
+            /// -----^^^ `stash_span`
+            /// |
+            /// `expr_span`
+            /// ```
+            /// `is_range_bound` is used to exclude arm guard suggestions in range pattern bounds.
+            fn emit_now(&self, stash_span: Span, expr_span: Span, is_range_bound: bool) {
+                self.parser.dcx().try_steal_modify_and_emit_err(
+                    stash_span,
+                    StashKey::ExprInPat,
+                    |err| {
+                        let sm = self.parser.psess.source_map();
+                        let stmt = self.stmt;
+                        let line_lo = sm.span_extend_to_line(stmt.span).shrink_to_lo();
+                        let indentation = sm.indentation_before(stmt.span).unwrap_or_default();
+                        let expr = self.parser.span_to_snippet(expr_span).unwrap();
+
+                        // Includes pre-pats (e.g. `&mut <err>`) in the diagnostic.
+                        err.span.replace(stash_span, expr_span);
+
+                        if let StmtKind::Let(local) = &stmt.kind {
+                            match &local.kind {
+                                // help: remove the `let`
+                                LocalKind::Decl | LocalKind::Init(_) => {
+                                    err.subdiagnostic(
+                                        UnexpectedExpressionInPatternSugg::RemoveLet {
+                                            let_span: local
+                                                .span
+                                                .shrink_to_lo()
+                                                .until(local.pat.span),
+                                            ty_span: local
+                                                .colon_sp
+                                                .map(|sp| sp.to(local.ty.as_ref().unwrap().span)),
+                                            has_initializer: !matches!(local.kind, LocalKind::Decl),
+                                        },
```

Comment:
Hmm, I'm not super convinced by this diagnostic. It has a lot of false positives. E.g., it suggests `1 + 2 = 3` for `let 1 + 2 = 3`. Which specific use case / user error did you want to target?

Reply:
I tried to handle `let`s but yeah that doesn't make munch sense, I removed the suggestions for `let _;` & `let _ = ...;`.

I initially kept them for `let arr[i] = ...`, but I'm not sure if that's common.